### PR TITLE
fix(source-google-ads): Fix schema loader for `custom_queries` streams 

### PR DESCRIPTION
**Pull request by tolik0**
## What

Fix several issues related to custom query support and memory performance:

- Refactor schema loading for custom queries to avoid low REST rate limits
- Switch the check stream from `campaign` to `customer` to reduce memory usage in Cloud environments
- Change endpoint for all streams from `search` to `searchStream`
- Add a streaming decoder to process `searchStream` responses incrementally, preventing large in-memory buffers

These improvements collectively reduce memory spikes during syncs and avoid hitting API rate limits on field metadata retrieval. Resolves: https://github.com/airbytehq/oncall/issues/9105

## How

- Refactored the schema loader for custom queries to use the official google-ads Python client instead of the REST metadata endpoint, avoiding its very low global rate limit (1,000 requests per app across all users)
- Updated the check stream to `customer`, which returns minimal rows while still validating API access and reducing memory pressure in Cloud environment
- Introduced a character-level streaming decoder that yields rows as soon as they are parsed from `searchStream`, significantly lowering peak memory usage for large streams

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Review by pnilan — APPROVED** [review]

**Comment by tolik0** [comment]
Regression tests:
https://github.com/airbytehq/airbyte/actions/runs/18567650835/job/52933066643
https://github.com/airbytehq/airbyte/actions/runs/18588661664/job/52998347909

**Comment by Aylr** [comment]
I'm looking forward to this nice fix being merged! ❤️

**Comment by geraldcor** [comment]
Ditto. This will likely unblock my team.

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/components.py [hunk 1]
```diff
@@ -939,3 +894,144 @@ def _validate_query(self, query: str):
...
+
+@dataclass
+class RowParseState:
+    collecting_row: bool = False
+    row_buffer: List[str] = field(default_factory=list)
+    row_nesting_depth: int = 0
+
+
+@dataclass
+class RowsStreamingDecoder(Decoder):
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/components.py]: nit: What does Row refers to here? Based on the code, I assume it is an entry within the `results` property, is that right? I'm not sure how the name could be better so can we document that at least?

**Review by maxi297 — APPROVED** [review]
The tests are clear and high quality. I like it! 

Can we add a bit more regarding edge cases? I'm thinking about:
* Different json formatting like separator being `(',', ':')`, different indents
* What types of error could happen during streaming a response and how it would impact this class? `StreamConsumedError` or `ChunkedEncodingError` maybe? 
* other error scenarios you can think of

The other concern I have is performance (recency bias here, not that I have specific signals to think this code is not performant).

All that being said, I'm preemptively approve the change as it looks good

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/components.py (hunk 1 above)
Comment (tolik0) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/components.py]: This is the terminology used by Google Ads to refer to records — I’ve renamed it to *records* for clarity.

Comment (maxi297) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/components.py]: If it is Google Ads terminology, I think it's fine. I just didn't know it. I'm fine with both actually

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/components.py [hunk 2]
```diff
@@ -939,3 +896,212 @@ def _validate_query(self, query: str):
...
+        if records_batch:
+            yield {"results": records_batch}
+
+    def _buffer_up_to_limit(self, response: requests.Response) -> Tuple[Union[bytes, Iterable[bytes]], bool]:
+        buf = bytearray()
+        response_stream = response.iter_content(chunk_size=self.CHUNK_SIZE)
+
+        while chunk := next(response_stream, None):
+            buf.extend(chunk)
+            if len(buf) > self.MAX_DIRECT_DECODE_BYTES:
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/components.py]: Because it is `>` and not `>=`, coes that mean that we will actually pull 25 Mb in memory instead of the threshold of 20 Mb described by `MAX_DIRECT_DECODE_BYTES`? I'm not actually opposed to pulling more in memory but I fear it might be confusing for the dev in case of a memory issue investigation

File: airbyte-integrations/connectors/source-google-ads/unit_tests/test_components.py [hunk 3]
```diff
@@ -288,3 +182,316 @@ def test_click_view_http_requester_returns_expected_request_body(self, config):
...
+    DecoderConfig(chunk_size=5 * 1024 * 1024, max_direct=20 * 1024 * 1024),  # huge chunk, big threshold → fast path
+]
+
+
+@pytest.fixture(params=CONFIGS, ids=lambda c: f"chunk={c.chunk_size}|max={c.max_direct}")
+def decoder(request):
+    cfg: DecoderConfig = request.param
+    d = GoogleAdsStreamingDecoder()
+    # monkey-patch per-test config
+    d.CHUNK_SIZE = cfg.chunk_size
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-google-ads/unit_tests/test_components.py]: nit: it feels like it could have been cleaner to have properties with default values. If at some point we want to parallelize some unit tests, we might struggle given those are shared across classes

**Review by maxi297 — APPROVED** [review]
LGTM! :shipit:

File: airbyte-integrations/connectors/source-google-ads/unit_tests/test_components.py (hunk 3 above)
Comment (tolik0) [reply airbyte-integrations/connectors/source-google-ads/unit_tests/test_components.py]: Fixed

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/components.py (hunk 2 above)
Comment (tolik0) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/components.py]: Yes, changed to >=
